In [ ]:
import torch
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
print(torch.__version__)

2.6.0+cu124


# Intro to tensors:


**torch.tensor**:
Tensor is a multi-dimensional matrix containing elements of a single data type.

A **scalar** is a tensor rank 0, so requires only 1 independent value associated with it

In [ ]:
# intro to tensors

scalar = torch.tensor(6)
print(scalar)
print(scalar.ndim)
print(scalar.shape)

tensor(6)
0
torch.Size([])


In [ ]:
torch.tensor([[1 , 2 , 3] , [ 4, 5 , 6]])

tensor([[1, 2, 3],
        [4, 5, 6]])

In [ ]:
torch.tensor(np.array([[1 , 2 , 3] , [ 4, 5 , 6]]))

tensor([[1, 2, 3],
        [4, 5, 6]])

torch.tensor() always copies data. If you have a Tensor data and just want to change its requires_grad flag, use requires_grad_() or detach() to avoid a copy. If you have a numpy array and want to avoid a copy, use torch.as_tensor().

In [ ]:
torch.zeros([2,4] )


tensor([[0., 0., 0., 0.],
        [0., 0., 0., 0.]])

In [ ]:
torch.zeros([2,4] , dtype = torch.int32)

tensor([[0, 0, 0, 0],
        [0, 0, 0, 0]], dtype=torch.int32)

In [ ]:
x = torch.tensor([[1,4,6] , [9,5,6]])
print(x[0][2])
x[1][1] = 3

x

tensor(6)


tensor([[1, 4, 6],
        [9, 3, 6]])

In [ ]:
x = torch.tensor([[1]])
print(x)
# get tensor as python int - using item()
x.item()


tensor([[1]])


1

In [ ]:
vector = torch.tensor([1,2,3])
print(vector)
print(vector.ndim)
print(vector.shape)

tensor([1, 2, 3])
1
torch.Size([3])


In [ ]:
# MATRIX
MATRIX = torch.tensor([[1,2,3] , [4,5,6]])
print(MATRIX)
print(MATRIX[0])
print(MATRIX[1])

tensor([[1, 2, 3],
        [4, 5, 6]])
tensor([1, 2, 3])
tensor([4, 5, 6])


In [ ]:
MATRIX.ndim


2

In [ ]:
MATRIX.shape

torch.Size([2, 3])

**scalar**:
 ndim: no of square brackets -> 0

**vector**:
shape : no of elements
ndim : no of square brackets -> 1

**matrix**:
shape : no of rows , col
ndim : no of square brackets -> 2

**tensor**:
ndim : no of square brackets -> 3 or more

shape : eg : 1 , 3, 3 means 1 3x3 shape tensor ( 1 dimension of 3x3)
very outer bracket matches with 1 , then next one matches with row , last with col

In [ ]:
TENSOR =  torch.tensor([[[1,4,5],
                         [3,4,5],
                         [6,7,4]]])
print(TENSOR)
print(TENSOR.ndim)
print(TENSOR.shape)

tensor([[[1, 4, 5],
         [3, 4, 5],
         [6, 7, 4]]])
3
torch.Size([1, 3, 3])


In [ ]:
# torch.size([1,3,3])
# indexes : torch.size([0,1,2])
# LOOK AT THE BRACKETS:
TENSOR[0]

tensor([[1, 4, 5],
        [3, 4, 5],
        [6, 7, 4]])

In [ ]:

TENSOR1 =  torch.tensor([[[[1,4,5],
                         [3,4,5],
                         [6,7,4]]]])
TENSOR1.ndim

4

In [ ]:
TENSOR1.shape

torch.Size([1, 1, 3, 3])

In [ ]:
TENSOR2 =  torch.tensor([[[[[1,4,5],
                         [3,4,5],
                         [6,7,4]]]]])
TENSOR2.shape



torch.Size([1, 1, 1, 3, 3])

*A tensor can represent any shape or size, or any combinations of number*





**usually:** a scalar and a vector is written by a lowercase and matrix or tensor is written by uppercase

### random tensors:

In [ ]:
# torch.rand(rows,cols)
random_tensor = torch.rand(3,4)
random_tensor

tensor([[0.0729, 0.2049, 0.3889, 0.8900],
        [0.5691, 0.5886, 0.0353, 0.2424],
        [0.8112, 0.2993, 0.9105, 0.0957]])

In [ ]:
random_tensor.ndim

2

In [ ]:
 random_tensor.shape

torch.Size([3, 4])

In [ ]:
random_TENSOR2 = torch.rand(1,3,4)
random_TENSOR2

tensor([[[0.1138, 0.4928, 0.2722, 0.5554],
         [0.2291, 0.5373, 0.8972, 0.9094],
         [0.0793, 0.5103, 0.3663, 0.0974]]])

In [ ]:
random_TENSOR2.shape

torch.Size([1, 3, 4])

In [ ]:
random_TENSOR2.ndim

3

In [ ]:
# create a random tensor with similar shape to an image tensor
random_image_size_tensor = torch.rand(size=(224,224,3) )
# height , width , color channel = 224 , 224 , 3
random_image_size_tensor.shape , random_image_size_tensor.ndim

(torch.Size([224, 224, 3]), 3)

ALMOST ANY DATA CAN BE REPRESENTED AS TENSORS , ONE OF THE COMMON WAYS OF REPRESENTING IMAGE DATA IS TO COLOR CHANNEL , HEIGHT , WIDTH OR HEIGHT , WIDTH , COLOR CHANNEL


### zeros and ones tensor

In [ ]:
# tensor of all zeroes:
zeroes = torch.zeros(size=(3,4))
zeroes

tensor([[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]])

In [ ]:
zeroes1 = torch.zeros((3,4))
zeroes1

tensor([[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]])

In [ ]:
random_value_tensor = torch.rand(3,4)
random_value_tensor

tensor([[0.5065, 0.2327, 0.5243, 0.5021],
        [0.1696, 0.1876, 0.5327, 0.3028],
        [0.4147, 0.2632, 0.1827, 0.5467]])

In [ ]:
zeroes1*random_value_tensor

tensor([[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]])

In [ ]:
# tensor of all ones:
ones = torch.ones((3,4) )
ones , ones.dtype

(tensor([[1., 1., 1., 1.],
         [1., 1., 1., 1.],
         [1., 1., 1., 1.]]),
 torch.float32)

# creating a range of tensors and tensors - like

In [ ]:
one_to_ten = torch.arange(1 , 11)
one_to_ten

tensor([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [ ]:
step_one_to_twenty = torch.arange(start = 1 , end = 20 , step = 4)
step_one_to_twenty

tensor([ 1,  5,  9, 13, 17])

In [ ]:
#  if we want to create a tensor full of zeros and has the same shape as one_to_ten we can use tensor like or zeros like
ten_zeros = torch.zeros_like(one_to_ten)
ten_zeros

tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

# Tensor datatypes

The default dtypes even if we specify as none are int64 for integer and float32 for float

In [ ]:
no_datatype_tensor = torch.tensor([3,6,8] , dtype = None)
no_datatype_tensor.dtype

torch.int64

In [ ]:
no_datatype_tensor = torch.tensor([3.0,6.0,8.0] , dtype = None)
no_datatype_tensor.dtype

torch.float32

In [ ]:
no_datatype_tensor = torch.tensor([3.0,6.0,8.0] , dtype = torch.float16)
no_datatype_tensor.dtype

torch.float16

Some of the important parameters of torch.tensor are dtype , device , requires_grad

you can calculate faster if numbers take less memory
*   single-precision floating-point :float32 (32 bits in computer memory)
*   half-precision floating point :float16

 the **precision** of a numerical quantity is a measure of the detail in which the quantity is expressed

 tensor dtypes is one of the 3 big errors youll run into with pytorch and deep learning:


1.   Tensor not right datatype
2.   Tensor not right shape
3.   Tensor not on right device

**Shape** :same concept as matrix multiplication

**device** : if one tensors on gpu for fast computing and another tensor is on cpu and you want to do some operations on them : then youll run into not right device error.
    device = "cuda"
    or device = "cpu"


**Requires_grad** : if you want pytorch to track gradients when pytorch goes through certain calculations

In [ ]:
float_32_tensor = torch.tensor([3.0 , 5.0 , 7.0] , dtype = torch.float32 ,
                               device = None , requires_grad = False)
float_32_tensor.device

device(type='cpu')

In [ ]:
# we can use either torch.float16 or torch.half
float_16_tensor = float_32_tensor.type(torch.half)
float_16_tensor

tensor([3., 5., 7.], dtype=torch.float16)

In [ ]:
different_datatype_mul = float_32_tensor*float_16_tensor
different_datatype_mul , different_datatype_mul.dtype

(tensor([ 9., 25., 49.]), torch.float32)

In [ ]:
int_32_tensor = torch.tensor([3,4,5] , dtype = torch.int32)
int_32_tensor

tensor([3, 4, 5], dtype=torch.int32)

In [ ]:
int_32_tensor * float_16_tensor

tensor([ 9., 20., 35.], dtype=torch.float16)

In [ ]:
some_tensor = torch.rand(4,5)
some_tensor

tensor([[0.6795, 0.8188, 0.7430, 0.2862, 0.5327],
        [0.9542, 0.9767, 0.0605, 0.7710, 0.2675],
        [0.3736, 0.1157, 0.7810, 0.7919, 0.8074],
        [0.4027, 0.5471, 0.8216, 0.1889, 0.8424]])

tensor.shape is an attribute and tensor.size() is a function and both give same values

In [ ]:
print(some_tensor)
print(f"Datatye of tensor : {some_tensor.dtype}")
print(f"Shape of tensor : {some_tensor.shape}")
print(f"Size of tensor : {some_tensor.size()}")
print(f"Device tensor is on : {some_tensor.device}")

tensor([[0.6795, 0.8188, 0.7430, 0.2862, 0.5327],
        [0.9542, 0.9767, 0.0605, 0.7710, 0.2675],
        [0.3736, 0.1157, 0.7810, 0.7919, 0.8074],
        [0.4027, 0.5471, 0.8216, 0.1889, 0.8424]])
Datatye of tensor : torch.float32
Shape of tensor : torch.Size([4, 5])
Size of tensor : torch.Size([4, 5])
Device tensor is on : cpu


### Manipulating tensors (tensor operations)

Tensor operations include ⁉


1.   Addition
2.   Subtraction
3.   Multiplication(element wise)
4.   Division
5.   Matrix multiplication



In [ ]:
tensor_op_addition = torch.tensor([1,2,3])
tensor_op_addition + 10

tensor([11, 12, 13])

In [ ]:
tensor_op_addition * 10

tensor([10, 20, 30])

In [ ]:
tensor_op_addition - 10

tensor([-9, -8, -7])

In [ ]:
# built in pytorch functions
torch.mul(tensor_op_addition , 20)


tensor([20, 40, 60])

**Matrix multiplication**

There are 2 ways of performing multiplication in deep learning:

*  Element-wise multiplication
*  Matrix multiplication

matrixmultiplication.xyz site is a good resource to learn matrix multiplication

In [ ]:
# element-wise multiplication
tensor_mul = torch.tensor([1,2,3])
print(tensor_mul , "*" , tensor_mul)
print(f"Tensor Element-wise multiplication : {tensor_mul * tensor_mul}")

tensor([1, 2, 3]) * tensor([1, 2, 3])
Tensor Element-wise multiplication : tensor([1, 4, 9])


You can either use matmul or @ like:


*   torch.matmul()
*   tensor @ tensor
*   torch.mm()

Important points about matrix multiplication are:


1.   the no of columns of first tensor should be equal to the no of rows of second tensor :

   eg :   
         tensor1's order = MxN
   
         tensor2's order = NxS

  
2.   The order of resulting matrix will be row of first tensor x col of second tensor that is:

          M x S




In [ ]:
# matrix multiplication - matmul
# here if the tensor is [1,2,3] then the matrix mul will be 1*1+2*2+3*3
%%time
torch.matmul(tensor_mul , tensor_mul)


CPU times: user 386 µs, sys: 0 ns, total: 386 µs
Wall time: 297 µs


tensor(14)

In [ ]:
%%time
value = 0
for i in range(len(tensor_mul)):
  value += tensor_mul[i]*tensor_mul[i]
print(value)

tensor(14)
CPU times: user 2.18 ms, sys: 15 µs, total: 2.19 ms
Wall time: 1.88 ms


We can also do this with the for loop but if pytorch already has a built in function for something complex then its better to use the built in function only because its fast


in the above example
*  Time with matmul function : 297 micro seconds
*  Time with loop : 1.88 milli seconds

(note that the time will change when we run it again)



In [ ]:
# example - if we try to multiply 2 tensors of order [3,2] , [3,2]
# that wont work - we get a shape error
torch.matmul(torch.rand(3,2) , torch.rand(3,2))

RuntimeError: mat1 and mat2 shapes cannot be multiplied (3x2 and 3x2)

In [ ]:
# but if we try to multiply two tensors of order [3,2] , [2,3] that will work
torch.matmul(torch.rand(3,2) , torch.rand(2,3))

tensor([[0.6136, 0.9717, 0.2482],
        [0.0970, 0.2015, 0.0324],
        [0.3059, 0.3656, 0.1407]])

In [ ]:
# another shape error example
tensor_A = torch.tensor([[1,2],
                         [3,4],
                         [5,6]])

tensor_B = torch.tensor([[6,7],
                         [4,7],
                         [3,9]])

torch.matmul(tensor_A, tensor_B)

RuntimeError: mat1 and mat2 shapes cannot be multiplied (3x2 and 3x2)

In [ ]:
# the above error was due to shapes , the col of first tensor is not the same as
# row of second tensor
tensor_A.shape , tensor_B.shape


(torch.Size([3, 2]), torch.Size([3, 2]))

To solve this we use **transpose**.


The **transpose of a matrix** is a new matrix formed by interchanging the rows and columns of the original matrix. If the original matrix is m x n, the transposed matrix will be n x m. The transpose of matrix A is often denoted as AT or A'.

In pytorch a transpose is written as -> tensor_name.T

In [ ]:
tensor_B

tensor([[6, 7],
        [4, 7],
        [3, 9]])

In [ ]:
tensor_B.T , tensor_B.T.shape
# now the shape of tensor b's transpose is 2,3 so that solves the issue

(tensor([[6, 4, 3],
         [7, 7, 9]]),
 torch.Size([2, 3]))

here in the above example ,

    tensor_A.shape = 3x2
    tensor_B.T.shape =  2x3

  col of tensor A is same as row of tensor b's transpose , and the resultant matrix will have the shape 3x3

In [ ]:
# now if we multiply transpose of tensor b and tensor a that will work
torch.matmul(tensor_A , tensor_B.T)

tensor([[20, 18, 21],
        [46, 40, 45],
        [72, 62, 69]])

In [ ]:
print(f"""Original shapes :
      tensor_A : {tensor_A.shape} ,
      tensor_B : {tensor_B.shape}""")


print(f"""\n B after transpose:
     {tensor_B.T.shape}""")

print(f"""\n Multiplying: {torch.matmul(tensor_A , tensor_B.T).shape}""")

print("output:\n")

output = torch.matmul(tensor_A , tensor_B.T)
print(output)

print(f"\n Output shape : {output.shape}")

Original shapes :
      tensor_A : torch.Size([3, 2]) ,
      tensor_B : torch.Size([3, 2])

 B after transpose:
     torch.Size([2, 3])

 Multiplying: torch.Size([3, 3])
output:

tensor([[20, 18, 21],
        [46, 40, 45],
        [72, 62, 69]])

 Output shape : torch.Size([3, 3])


# Tensor Aggregation : min , max , mean , sum

In [ ]:
x = torch.arange(1 , 100 , 10)
x , x.dtype

(tensor([ 1, 11, 21, 31, 41, 51, 61, 71, 81, 91]), torch.int64)

In [ ]:
torch.min(x) , x.min()

(tensor(1), tensor(1))

In [ ]:
torch.max(x) , x.max()

(tensor(91), tensor(91))

A most common type of pytorch error : datatype error:

for torch.mean function input datatype must be a floating pt or complex

pytorch document does not cover this detail about torch.mean method

In [ ]:
# example of a datatype error
torch.mean(x)

RuntimeError: mean(): could not infer output dtype. Input dtype must be either a floating point or complex dtype. Got: Long

In [ ]:
# we can convert x into float32 or any other accepted dtype
torch.mean(x.type(torch.float32))

tensor(46.)

In [ ]:
torch.sum(x) , x.sum()

(tensor(460), tensor(460))

# finding the positional min and max : argmin , argmax

In [ ]:
x

tensor([ 1, 11, 21, 31, 41, 51, 61, 71, 81, 91])

**Argmin()**:

Find the position in tensor that has the minimum value with argmin() -> **returns index position** of target tensor where the minimum value occurs

If there are multiple minimal values then the indices of the first minimal value are returned.

In [ ]:
# RETURNS INDEX
torch.argmin(x) , x.argmin()

(tensor(0), tensor(0))

**Argmax()**:
Find the position in tensor that has the max value with argmax() -> it also returns **index**

In [ ]:
torch.argmax(x) , x.argmax()

(tensor(9), tensor(9))

# Reshaping , viewing , stacking , squeezing and unsqueezing



*   **Reshaping** : Reshapes an input tensor to a defined shape . Returns a tensor with the same data and number of elements as input, but with the specified shape.

*   **View** : Return a view of an input tensor of a certain shape but keep the same memory as the original tensor
*   **Stacking** : Combines multiple tensors on top of each other (vstack-> vertical stack ) or side by side (hstack -> horizontal stack)
*   **Squeeze** : Removes all `1` dimensions from a tensor
*   **Unsqueeze** : Adds a `1` dimension to a target tensor
*   **Permute** : Return a view of the input with dimensions permuted (swapped) in a certain way







In [ ]:
y = torch.arange(1. , 10.)
y , y.shape

(tensor([1., 2., 3., 4., 5., 6., 7., 8., 9.]), torch.Size([9]))

In [ ]:
# add an extra dimension
x_reshaped = x.reshape(1,10)
x_reshaped

tensor([[ 1, 11, 21, 31, 41, 51, 61, 71, 81, 91]])